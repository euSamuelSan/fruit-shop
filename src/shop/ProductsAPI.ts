import apple from '../assets/images/apple.png'
import banana from '../assets/images/banana.png'
import mango from '../assets/images/mango.png'
import pear from '../assets/images/pear.png'
import pineapple from '../assets/images/pineapple.png'

export type Fruit = {
    id: string
    name: string
    price: number
    image: string
    measurementUnit: string
}

type GetAllResponse = {
    data: Fruit[]
}

type GetResponse = {
    data: Fruit
}

export const getAllFruits = async (): Promise<GetAllResponse | void> => {
    try {
        return Promise.resolve({ data: fruitsMock })
    } catch (err) {
        console.error(err)
    }
}

export const getFruitById = async (id: string): Promise<GetResponse | void> => {
    try {
        const item = fruitsMock.find(fruit => fruit.id === id)

        return item
            ? Promise.resolve({ data: item })
            : Promise.reject('ID informado não encontrado.')
    } catch (err) {
        console.error(err)
    }
}

export const fruitsMock: Fruit[] = [
    {
        id: '1',
        name: 'Maçã',
        price: 5.56,
        image: apple,
        measurementUnit: 'KG',
    },
    {
        id: '2',
        name: 'Pêra',
        price: 7.22,
        image: pear,
        measurementUnit: 'KG',
    },
    {
        id: '3',
        name: 'Banana',
        price: 2.5,
        image: banana,
        measurementUnit: 'KG',
    },
    {
        id: '4',
        name: 'Abacaxi',
        price: 6.65,
        image: pineapple,
        measurementUnit: 'Un',
    },
    {
        id: '5',
        name: 'Manga',
        price: 4.44,
        image: mango,
        measurementUnit: 'KG',
    },
]
