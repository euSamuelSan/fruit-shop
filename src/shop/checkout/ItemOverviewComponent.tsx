import React from 'react'
import styled from 'styled-components'
import { AiFillPlusCircle, AiFillMinusCircle } from 'react-icons/ai'

import { standardizePrice } from '../../commons/StringUtils'
import Colors from '../../commons/Colors'
import { useUserContext } from '../../context/UserContext'
import { Fruit } from '../ProductsAPI'

export type ItemOverviewComponentProps = {
    product: Fruit & { quantity: number }
    index: number
}
export default function ItemOverviewComponent({
    product,
    index,
}: ItemOverviewComponentProps): JSX.Element {
    const { onChangeCartItems } = useUserContext()

    const addItem = () => onChangeCartItems('add', product.id)
    const removeItem = () => onChangeCartItems('subtract', product.id)

    const renderActions = () => {
        return (
            <ActionsContainer>
                <Action onClick={() => removeItem()}>
                    <AiFillMinusCircle
                        size={20}
                        title="Remover"
                        color={Colors.blueGrey}
                    />
                </Action>
                <QuantityText>
                    {product.quantity} {product.measurementUnit}
                </QuantityText>
                <Action onClick={() => addItem()}>
                    <AiFillPlusCircle
                        size={20}
                        title="Adicionar"
                        color={Colors.blueGrey}
                    />
                </Action>
            </ActionsContainer>
        )
    }
    return (
        <Container index={index}>
            <Image src={product.image} alt={product.name} />
            <Text>{product.name}</Text>
            <Text>
                <b>{standardizePrice(product.price * product.quantity)}</b>(
                {standardizePrice(product.price)} por {product.measurementUnit})
            </Text>
            {renderActions()}
        </Container>
    )
}

const Container = styled.div<Partial<ItemOverviewComponentProps>>`
    display: grid;
    grid-template-columns: 0.4fr 0.7fr 1.9fr 1fr;
    height: 40px;
    padding: 0 10px;

    @media (max-width: 360px) {
        grid-template-columns: 0.8fr 2fr 1.2fr;
        padding: 0 5px;
    }

    background-color: ${props =>
        typeof props.index === 'number' && props.index % 2 === 0
            ? Colors.lightBlueGreyFour
            : Colors.white};
`

const Image = styled.img`
    max-width: 35px;
    max-height: 35px;

    display: block;
    margin: auto;

    @media (max-width: 500px) {
        max-width: 30px;
        max-height: 30px;
    }

    @media (max-width: 360px) {
        display: none;
    }
`

const Text = styled.p`
    color: ${Colors.black};
    margin: 0;
    line-height: 38px;
    font-size: 14px;

    @media (max-width: 500px) {
        font-size: 12px;
    }
`

const QuantityText = styled(Text)`
    font-size: 12px;
    @media (max-width: 500px) {
        font-size: 10px;
    }
`

const ActionsContainer = styled.div`
    display: flex;
    align-items: center;
`

const Action = styled.button`
    display: flex;
    border: none;
    background: none;
    cursor: pointer;
    justify-content: center;
    align-content: center;
    align-items: center;
    border-radius: 25px;

    width: 34px;
    height: 34px;

    @media (max-width: 500px) {
        width: 30px;
        height: 30px;
    }

    @media (max-width: 360px) {
        width: 27px;
        height: 27px;
    }

    &:hover {
        background-color: rgb(0, 0, 0, 0.1);
        transition: background-color 0.2s;
    }
`
