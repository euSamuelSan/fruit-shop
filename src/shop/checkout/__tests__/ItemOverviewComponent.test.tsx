import { render, RenderResult } from '@testing-library/react'
import { fruitsMock } from '../../ProductsAPI'

import ItemOverviewComponent, {
    ItemOverviewComponentProps,
} from '../ItemOverviewComponent'

describe('ItemOverviewComponent', () => {
    const getProps = (): ItemOverviewComponentProps => ({
        product: {
            ...fruitsMock[0],
            quantity: 1,
        },
        index: 1,
    })

    const renderItemOverviewComponent = (props = getProps()): RenderResult => {
        return render(<ItemOverviewComponent {...props} />)
    }

    it('Deve desenhar o componente corretamente', () => {
        renderItemOverviewComponent()

        expect(document.body).toMatchSnapshot()
    })
})
