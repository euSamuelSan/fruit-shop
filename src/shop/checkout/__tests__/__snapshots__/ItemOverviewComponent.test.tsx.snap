// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemOverviewComponent Deve desenhar o componente corretamente 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ jCrKuC"
    >
      <img
        alt="Maçã"
        class="sc-gsnTZi dLiuVt"
        src="apple.png"
      />
      <p
        class="sc-dkzDqf cpOprJ"
      >
        Maçã
      </p>
      <p
        class="sc-dkzDqf cpOprJ"
      >
        <b>
          R$ 5,56
        </b>
        (
        R$ 5,56
         por 
        KG
        )
      </p>
      <div
        class="sc-eCYdqJ cWIIhL"
      >
        <button
          class="sc-jSMfEi fSwEsX"
        >
          <svg
            color="#8e8e93"
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(142, 142, 147);"
            viewBox="0 0 1024 1024"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Remover
            </title>
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm192 472c0 4.4-3.6 8-8 8H328c-4.4 0-8-3.6-8-8v-48c0-4.4 3.6-8 8-8h368c4.4 0 8 3.6 8 8v48z"
            />
          </svg>
        </button>
        <p
          class="sc-dkzDqf sc-hKMtZM cpOprJ jjQKFz"
        >
          1
           
          KG
        </p>
        <button
          class="sc-jSMfEi fSwEsX"
        >
          <svg
            color="#8e8e93"
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(142, 142, 147);"
            viewBox="0 0 1024 1024"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Adicionar
            </title>
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm192 472c0 4.4-3.6 8-8 8H544v152c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V544H328c-4.4 0-8-3.6-8-8v-48c0-4.4 3.6-8 8-8h152V328c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v152h152c4.4 0 8 3.6 8 8v48z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</body>
`;
